src/mds/SnapRealm.cc

```diff
@@ -345,8 +345,9 @@ void SnapRealm::split_at(SnapRealm *child)
     }
   }
 
-  // split inodes_with_caps
-  for (auto p = inodes_with_caps.begin(); !p.end(); ) {
```

Comment:
so what about `inodes_with_caps`, shouldn't they get treated here as well?

Reply:
`inodes` is a superset of `inodes_with_caps`. `CInode::move_to_realm` will also move itself to the correct `inodes_with_caps` list.

Reply:
This is not obvious. I understand how this is handled now, but maybe there's a better way? having two lists where one is a subset of the other could be a source of non-obvious bugs.

Do we need the `inodes_with_caps` at all if we have the superset? would it be wasteful to scan the `inodes` in places where `inodes_with_caps` was used until today?

Reply:
These lists are pretty cheap as elist only introduces an overhead in the size of the CInode.

There is a larger issue that prevents my attempt to merge the two. CInode::inodes_with_caps is apparently used in two places (as hinted by the comment I added in CInode.h). I don't like that and it should be fixed. I wrote up a ticket here:

https://tracker.ceph.com/issues/64008

Thanks for the nudge.

Reply:
Further thinking about this, I'd like to question why we want to add and maintain this inode list if we could also just traverse the loaded inode tree via the available inode and cdir and cdir's loaded dentries? It will be less performant than a dedicated inode list, but it should be that slow either. We already have the tree available and will just add a couple of indirections to jump over all loaded subtree nodes. 
What is preventing us from traversing the already available in-memory tree directly?

Reply:
yeah, but you had to introduce a new list for that now. I am not sure it's reasonable to rely on the snap realm if all we need is to enumerate all loaded inodes. While it's more performant to traverse a list, I don't think we'll notice the difference if we traverse an the in-memory tree structure. However, the downsides are

* binding the implementation to SnapRealm without the real need is both confusing and limiting (we can't support non-snap realm enabled directories and we must be careful only to use childless snap realms)
* introducing another list that's a superset of the existing list causes awkward cases in the code and isn't obvious

Reply:
> yeah, but you had to introduce a new list for that now. I am not sure it's reasonable to rely on the snap realm if all we need is to enumerate all loaded inodes. While it's more performant to traverse a list, I don't think we'll notice the difference if we traverse an the in-memory tree structure. However, the downsides are

The performance difference is probably not significant.
 
>     * binding the implementation to SnapRealm without the real need is both confusing and limiting (we can't support non-snap realm enabled directories and we must be careful only to use childless snap realms)
> 
>     * introducing another list that's a superset of the existing list causes awkward cases in the code and isn't obvious

My preference would be to be deliberately restrictive in the initial implementation. A few years ago, I worked with Zheng to have the "ceph.dir.subvolume" flag created to restrict nested snapshots within the subvolumes created by the volumes plugin; this was to avoid any potential bugs and performance issues. (Despite this, we've had bug even with "trivial" snap realm configurations: see the famous dentry/cinode "first" metadata corruption.) I believe there may be similar issues here with respect to split subtrees, active exports/imports. The new distributed ephemeral pinning we now have on by default [1] should limit the scope of those issues within this context.

So if there's time, I would like to revisit the use of the list but I want to emphasize that it can be easily reverted later if we want to make the quiesce protocol more generally applicable.

[1] https://github.com/rook/rook/pull/12477